datastore/warehouse/warehouse.go

```diff
@@ -0,0 +1,278 @@
+
+	shutdown       chan struct{}
+	runningWorkers *structs.TimeoutWaitGroup
+
+	m WarehouseMetrics
+}
+
+func NewWarehouse(logger log.Logger, bufSize int) *Warehouse {
+	wh := &Warehouse{logger: logger.WithField("service", "warehouse"), requests: make(chan StoreRequest), runningWorkers: structs.NewTimeoutWaitGroup()} // channel must be unbuffered, for not accepting requests that will not be handled
+	wh.initMetrics()
+	return wh
+}
+
+func (s *Warehouse) RunParallel(ctx context.Context, datadir string, numWorkers int) error {
+	for i := 0; i < numWorkers; i++ {
+		go s.Run(ctx, datadir, i)
+	}
+
+	return nil
+}
+
+func (s *Warehouse) Run(ctx context.Context, datadir string, id int) {
+	s.runningWorkers.Add(1)
+	defer s.runningWorkers.Done()
+
+	logger := s.logger.With(log.F{
+		"workerId": id,
+		"datadir":  datadir,
+	})
+
+	logger.Info("started")
+	defer logger.Info("stopped")
+
+	w := newWorker(id, datadir, logger)
+	ticker := time.NewTicker(filesPruneTick)
+	for {
+		select {
+		case req := <-s.requests:
+			s.handleRequest(ctx, logger, w, req)
+			continue
+		case <-ticker.C:
+			w.closeIdleFiles()
+			continue
+		case <-ctx.Done():
+			w.closeAllFiles()
+			return
+		case <-s.shutdown:
+			w.closeAllFiles()
+			return
+		}
+	}
+}
+
+func (s *Warehouse) handleRequest(ctx context.Context, logger log.Logger, w *worker, req StoreRequest) {
+	file, err := w.getOrCreateFile(req)
+	if err == nil {
+		err = writeToFile(file, req)
+		if err != nil {
+			logger.WithError(err).With(req).Error("failed to write")
+
+			file.Write([]byte("\n"))
+			w.closeFile(file)
+			logger.WithField("filename", file.Name()).Debug("corrupted file closed")
+		}
+	} else {
+		logger.WithError(err).Error("failed to get/create file: %w", err)
+	}
+
+	if err == nil {
+		s.m.Writes.WithLabelValues(toString(req.DataType)).Add(1)
+	} else {
+		s.m.FailedWrites.WithLabelValues(toString(req.DataType)).Add(1)
+	}
+
+	select {
+	case req.err <- err: // if does not block, means it is buffered (1) channel
+	default:
+	}
+
+	select {
+	case req.err <- err:
+		return
+	case <-s.shutdown:
+		logger.WithError(ErrClosed).Error("failed to export request")
+		return
+	case <-ctx.Done():
+		logger.WithError(ctx.Err()).Error("failed to export request")
+		return
+	}
+}
+
+func (s *Warehouse) Close(ctx context.Context) {
+	close(s.shutdown) // prevent new requests from being added
+
+	// wait for inflight requests to complete
+	select {
+	case <-s.runningWorkers.C():
+	case <-ctx.Done():
+	}
+}
+
+func (s *Warehouse) Store(ctx context.Context, req StoreRequest) error {
+	if err := s.StoreAsync(ctx, req); err != nil {
+		return err
+	}
+
+	// wait for response
+	select {
+	case err := <-req.err:
+		return err
+	case <-s.shutdown:
+		return ErrClosed
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+func (s *Warehouse) StoreAsync(ctx context.Context, req StoreRequest) error {
+	// check if service is closed
+	select {
+	case <-s.shutdown:
+		return ErrClosed
+	default:
+	}
```

Comment:
if we offloaded something to get asynchronously stored, we probably don't like to loose it :) I think the strategy should be different. So, wait until everything that was requested was finished, "however long it would take" and then finish. So I would remove this and also 154-155